external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala

```diff
@@ -1374,4 +1377,182 @@ class AvroSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
+
+    assert(row.getStruct(0, 4).get(0, IntegerType) == null)
+    assert(row.getStruct(0, 4).get(1, FloatType) == null)
+    assert(row.getStruct(0, 4).get(2, StringType) == null)
+    assert(row.getStruct(0, 4).getStruct(3, 2) == null)
+    assert(record == recordFromRow)
+
+    record.put(0, 1)
+    row = expressionEncoder.toRow(record)
+    recordFromRow = expressionEncoder.resolveAndBind().fromRow(row)
+
+    assert(row.getStruct(0, 4).get(1, FloatType) == null)
+    assert(row.getStruct(0, 4).get(2, StringType) == null)
+    assert(row.getStruct(0, 4).getStruct(3, 2) == null)
+    assert(record == recordFromRow)
+
+    record.put(0, 1F)
+    row = expressionEncoder.toRow(record)
+    recordFromRow = expressionEncoder.resolveAndBind().fromRow(row)
+
+    assert(row.getStruct(0, 4).get(0, IntegerType) == null)
+    assert(row.getStruct(0, 4).get(2, StringType) == null)
+    assert(row.getStruct(0, 4).getStruct(3, 2) == null)
+    assert(record == recordFromRow)
+
+    record.put(0, "bar")
+    row = expressionEncoder.toRow(record)
+    recordFromRow = expressionEncoder.resolveAndBind().fromRow(row)
+
+    assert(row.getStruct(0, 4).get(0, IntegerType) == null)
+    assert(row.getStruct(0, 4).get(1, FloatType) == null)
+    assert(row.getStruct(0, 4).getStruct(3, 2) == null)
+    assert(record == recordFromRow)
+
+    record.put(0, new GenericRecordBuilder(nested).build())
+    row = expressionEncoder.toRow(record)
+    recordFromRow = expressionEncoder.resolveAndBind().fromRow(row)
+
+    assert(row.getStruct(0, 4).get(0, IntegerType) == null)
+    assert(row.getStruct(0, 4).get(1, FloatType) == null)
+    assert(row.getStruct(0, 4).get(2, StringType) == null)
+    assert(record == recordFromRow)
+  }
+
+  test("create Dataset from GenericRecord") {
+    // need a spark context with kryo as serializer
+    val conf = new SparkConf()
+      .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+      .set("spark.driver.allowMultipleContexts", "true")
+      .set("spark.master", "local[2]")
+      .set("spark.app.name", "AvroSuite")
+    val context = new SparkContext(conf)
+
+    val schema: Schema =
+      SchemaBuilder
+        .record("GenericRecordTest")
+        .namespace("com.databricks.spark.avro")
+        .fields()
+        .requiredString("field1")
+        .name("enumVal").`type`().enumeration("letters").symbols("a", "b", "c").enumDefault("a")
+        .name("fixedVal").`type`().fixed("MD5").size(16).fixedDefault(ByteBuffer.allocate(16))
+        .endRecord()
+
+    implicit val enc = AvroEncoder.of[GenericData.Record](schema)
+
+    val genericRecords = (1 to 10) map { i =>
+      new GenericRecordBuilder(schema)
+        .set("field1", "field-" + i)
+        .build()
+    }
+
+    val rdd: RDD[GenericData.Record] = context.parallelize(genericRecords)
+    val ds = rdd.toDS()
+    assert(ds.count() == genericRecords.size)
+    context.stop()
+  }
 }
```

Comment:
The above tests are all for `GenericRecord` Avro classes. It might be good to generate an Avro class having a schema similar to the GenericRecord described [above](https://github.com/apache/spark/pull/22878/files#diff-9364b0610f92b3cc35a4bc43a80751bfR1386), so that we can test an instance extending `SpecificRecord` (which will probably be the most commonly used Avro class for the encoder).

There was one such [class](https://github.com/databricks/spark-avro/pull/217/files#diff-6088df231dbde4904100296c8a90fe93) in the Spark-Avro project, but I can understand why it may not have been copied over in this PR.

Reply:
Yep, actually I test the cases you mentioned self but its need to add lots of generation code by avro. Moreover, IIUC, the testing of `SpecificRecord` just test one more logic of `avroClass.getMethod("getClassSchema")`, I just think no need to add those generation code for this test. If we really want to achieve this maybe add a little simple specific record example based on existing `test.avsc`? Or if we just want to show the usage, maybe add corresponding document is enought. WDYT :)

Reply:
Yeah, I'm comfortable with that. The rest of the encoder code path would be the same.